node-launchpad/src/system.rs

```diff
@@ -0,0 +1,148 @@
+        }
+    }
+    // If no matching disk is found, return an empty string or handle the error as needed
+    (String::new(), String::new())
+}
+
+// Checks if the given path (a drive) is read-only
+fn is_read_only<P: AsRef<Path>>(path: P) -> bool {
+    let test_file_path = path.as_ref().join("lauchpad_test_write_permission.tmp");
+
+    // Try to create and write to a temporary file
+    let result = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(&test_file_path)
+        .and_then(|mut file| file.write_all(b"test"));
+
+    match result {
+        Ok(_) => {
+            // Clean up the test file if write was successful
+            let _ = std::fs::remove_file(test_file_path);
+            false
+        }
+        Err(err) => {
+            // Check if the error is due to a read-only file system
+            err.kind() == std::io::ErrorKind::PermissionDenied
+        }
+    }
+}
+
+// Gets a list of drives and their available space
+pub fn get_list_of_drives_and_available_space() -> Vec<(String, String, u64)> {
+    // Create a new System instance
+    let disks = Disks::new_with_refreshed_list();
+
+    // Get the list of disks
+    let mut drives: Vec<(String, String, u64)> = Vec::new();
+    for disk in disks.list() {
+        // Check if the disk is already in the list
+        let disk_info = (
+            disk.name()
+                .to_string_lossy()
+                .into_owned()
+                .trim()
+                .to_string(),
+            disk.mount_point()
+                .to_string_lossy()
+                .into_owned()
+                .trim()
+                .to_string(),
+            disk.available_space(),
+        );
+        // We don't check for write permission on removable drives
+        if !disk.is_removable() {
+            // Check if the disk is read-only and skip it
+            if is_read_only(disk.mount_point()) {
+                continue;
+            }
+        }
+        if !drives.contains(&disk_info) {
+            drives.push(disk_info);
+        }
+    }
+    debug!("Drives detected: {:?}", drives);
+    drives
+}
+
+// Opens a folder in the file explorer
+pub fn open_folder(path: &str) -> std::io::Result<()> {
+    if Path::new(path).exists() {
+        #[cfg(target_os = "macos")]
+        Command::new("open").arg(path).spawn()?.wait()?;
+        #[cfg(target_os = "windows")]
+        Command::new("explorer").arg(path).spawn()?.wait()?;
+        #[cfg(target_os = "linux")]
+        Command::new("xdg-open").arg(path).spawn()?.wait()?;
+    } else {
+        error!("Path does not exist: {}", path);
+    }
+    Ok(())
+}
+
+#[cfg(unix)]
+pub fn get_mount_point() -> PathBuf {
+    PathBuf::from("/")
+}
+#[cfg(windows)]
+pub fn get_mount_point() -> PathBuf {
+    PathBuf::from("C:\\")
+}
+
+// Gets available disk space in bytes for the given mountpoint
+pub fn get_available_space_b(storage_mountpoint: String) -> Result<usize> {
```

Comment:
Just a tiny nit pick, you can avoid this tbh. 1) But we can just pass `&str` as argument here OR 2) inside the functions, instead of using `storage_mountpoint.clone()`, we could also just use the references `&storage_mountpoint`. I think the compiler would optimize these `clones` out, but just to be explicit, we could always try using references whenever possible.

Reply:
Thanks Roland. I've switched to `PathBuf` and used reference to avoid `clone()` 👍